### PR TITLE
faction camp: menial labor uses the zone autosort feature

### LINKED ISSUE #25397: NPCs ignore 'No NPC Pickup' zone when completing Menial Labor task at camp
**Describe the bug**  
NPCs assigned to the 'Menial Labor' task in NPC camps will collect items stored in an assigned 'No NPC Pickup' zone. Not sure if this was intended or not but it can be annoying if you have a preexisting base nearby an NPC camp and they clear out your whole base in one go.

Build 7841

### LINKED ISSUE #28880: Faction camp "drop points reset" little bug
**Describe the bug**  
When you press reset drop points, it asks you if you want to reset every point. But if you say "no" it will reset it to a very far point on the map. So its only possible to set everything up in a single run, or don't touch that feature at all. Since it doesn't shows you how do you set your stuff previously, its quite hard to remember all old placements.

**Expected behavior**  
A possibility to change only a few drop points instead of all.

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "faction camp: menial labor uses the zone autosort feature"```

#### Purpose of change
Fixes #25397
Fixes #28880
Fix faction camp menial labor sorting by making it use the existing zone autosort feature.  As a side effect, add the infrastructure for NPCs performing player activities on an ongoing basis. 

#### Describe the solution
Add a function to clzones to determine if there are any loot zones nearby.

Alter the npc `*_destination` functions so it is clear that deal with omt destinations.

Add infrastructure to npcmove to let friendly NPCs perform player activities using the NPC_MISSION_ACTIVITY mission.  NPCs gain bonus moves from elapsed time during `npc::on_load()` if they have an activity, and lose those bonus moves when the activity is complete and their activity backlog is empty

Update do_turn_move_loot to support NPCs instead of the player performing the activity.  Handle a weird NPC only corner case where the NPC is still active but not in the bounds of map by unloading them.

Change validate_sort_points to check if there is an unsorted loot zone, sorted loot zone, and camp food zone nearby.  Change the Distribute Food pseudo mission to take food from the camp food zone.  Change the Menial Labor mission to have the NPC perform the loot sorting activity using the infrastructure set up earlier in this PR.

#### Describe alternatives you've considered
Leaving the NPC in the overmap buffer and copy-pasting the do_activity_move_turn code into menial_return is easier but not the best solution.

### CODE REVIEW

**Comment by mlangsdorf** [comment]
Current status: the NPC does some sorting, taking up time and freezing PC activity.  I don't think that's right.

I think the issue is that `activity_on_turn_move_loot` is doing too much per cycle.  I want it to do exactly 1 turn's worth of work per invocation, and I don't think it's doing that.  I think it's sorting one entire tile and then getting invoked again.

File: src/activity_item_handling.cpp [hunk 1]
```diff
@@ -1093,23 +1095,27 @@ void activity_on_turn_move_loot( player_activity &, player &p )
                     }
                     // check free space at destination
                     if( free_space >= ( *it )->volume() ) {
-                        // before we move any item, check if player is at or adjacent to the loot source tile
+                        // before we move any item, check if player is at or
+                        // djacent to the loot source tile
```
Comment (ifreund) [inline src/activity_item_handling.cpp]: looks like you lost an "a" here, and an "s" in the next two comment changes

**Comment by ifreund** [comment]
The problem is that `move_item` [here](https://github.com/CleverRaven/Cataclysm-DDA/pull/28571/files#diff-999040b151800caa2f76bd43d4b4d43dL1130) modifies the player's moves not the npc's, so [this check](https://github.com/CleverRaven/Cataclysm-DDA/pull/28571/files#diff-999040b151800caa2f76bd43d4b4d43dL1138) never succeeds 

Also, letting npcs sort may cause some strange behavior if the player defines new zones during sorting.  I'm pretty sure that because of the caching that any new `LOOT_UNSORTED` will simply be ignored, and new destination zones will be properly sorted to if they are closest, but it would be good to investigate.

File: src/activity_item_handling.cpp (hunk 1 above)
Comment (mlangsdorf) [reply src/activity_item_handling.cpp]: Thanks, fixed.

**Comment by mlangsdorf** [comment]
@ifreund if you have a chance, please review or playtest.  thanks!

File: src/faction_camp.cpp [hunk 2]
```diff
@@ -637,9 +625,9 @@ void basecamp::get_available_missions( mission_data &mission_key )
 
         entry = string_format( _( "Notes:\n"
                                   "Distribute food to your follower and fill you larders.  Place "
-                                  "the food you wish to distribute on the companion food sort "
-                                  "point.  By default, that is opposite the tent door between "
-                                  "the manager and wall.\n \n"
+                                  "the food you wish to distribute in the camp food zone.  You "
+                                  "must create a camp food zonei near the camp using the zone "
```
Comment (ifreund) [inline src/faction_camp.cpp]: "zonei" -> zone

File: src/faction_camp.cpp [hunk 3]
```diff
@@ -1199,8 +1186,9 @@ bool basecamp::handle_mission( const std::string &miss_id, const std::string &mi
     }
 
     if( miss_id == "Menial Labor" ) {
-        start_mission( "_faction_camp_menial", 3_hours, true,
-                       _( "departs to dig ditches and scrub toilets..." ), false, {}, "", 0 );
+        //start_mission( p, "_faction_camp_menial", 3_hours, true,
```
Comment (ifreund) [inline src/faction_camp.cpp]: remove commented code

File: src/faction_camp.cpp [hunk 4]
```diff
@@ -2078,70 +2084,138 @@ bool basecamp::menial_return()
     if( comp == nullptr ) {
         return false;
     }
+#if 0
```
Comment (ifreund) [inline src/faction_camp.cpp]: probably should remove this code too

File: src/faction_camp.cpp [hunk 5]
```diff
@@ -3101,59 +3175,24 @@ std::vector<item *> basecamp::give_equipment( std::vector<item *> equipment,
...
-        sort_points.clear();
-        sort_points.insert( sort_points.end(), new_pts.begin(), new_pts.end() );
-        return true;
-    }
-    if( query_yn( _( "Revert to default points?" ) ) ) {
-        return set_sort_points( true, false );
-    }
-    return false;
+    popup( _( "Sorting zones have changed.  Please create some sorting zones.  "
+              "You must create a camp food zone for your camp." ) );
```
Comment (ifreund) [inline src/faction_camp.cpp]: This should mention the other required zones as well.

File: src/npcmove.cpp [hunk 6]
```diff
@@ -2375,6 +2398,26 @@ bool npc::do_pulp()
...
 }
 
+bool npc::do_player_activity()
+{
+    int old_moves = moves;
+    while( moves > 0 && activity ) {
+        activity.do_turn( *this );
+    }
+    /* if the activity is finished, grab any backlog or change the mission */
+    if( !has_destination() && !activity ) {
```
Comment (ifreund) [inline src/npcmove.cpp]: Should this be `has_destination_activity()` instead of `has_destination()`?

Comment (mlangsdorf) [reply src/npcmove.cpp]: You'd think that, wouldn't you?  I sure did.

has_destination_activity() only returns true when you have a destination_activity and you've reached your destination for automove.  It's used to detect that you can resume an activity that was interrupted for movement.

This is a check that an activity has completed entirely, so no activity and no autoroute movement is the right check.

**Comment by ifreund** [comment]
Shoot, somehow it took me until now to realize that this actually totally breaks caching if the player decides to remove items from the unsorted zone.  What we need to do is clear the cache then restart the activity one extra time the first time we think we are done.  There's a low chance that could still leave some items unsorted, but I think that's not a huge deal.

**Comment by mlangsdorf** [comment]
I will address your caching concerns tomorrow when I am slightly less brain-fried.

**Comment by ifreund** [comment]
Alright just to flesh out what I said about caching a bit: 
* Adding items to a zone while sorting is no problem, as items are added to the top and sorting checks from bottom to top in order.  
* Removing items that had been cached as unsortable will cause an equal number of sortable items on that tile to be skipped.
* The most complete fix is to always restart the activity using `mgr.start_sort()` and `mgr.end_sort()` to clear the cache unless we iterated through all the items and reached the end of the sort without restarting the activity.  If that happens then all remaining items are unsortable and we have completed the activity.

Hopefully that all makes sense.

**Comment by mlangsdorf** [comment]
jenkins rebuild

**Comment by mlangsdorf** [comment]
Resolved the caching issue by calling end_sort() on any exit from the function.

Added logic to handle the NPC leaving the reality bubble by unloading them and then giving them a lot of bonus moves when they're loaded back into the reality bubble.  It seems to work in my tests but could use more testing.

**Pull request merged**